agera/src/main/java/com/google/android/agera/FunctionCompilerStates.java

```diff
@@ -144,4 +152,87 @@
     @NonNull
     Function<TFrom, TPrevList> thenSort(@NonNull Comparator<TPrev> comparator);
   }
+
+  /**
+   * Compiler state allowing to specify how the {@link Function} should modify {@link Result}s.
+   */
+  interface FResult<TPrev, TFrom> {
+
+    /**
+     * Adds a {@link Function} to the behavior chain to map the item into a new type. If the result
+     * failed, then we return {@code sameFailure}.
+     *
+     * @param function the function to apply to the item to create a new result
+     */
+    @NonNull
+    <TTo> FResult<TTo, TFrom> applyIfSucceeded(@NonNull Function<? super TPrev, TTo> function);
+
+    /**
+     * Adds a {@link Function} to the end of the behavior chain to map the item into a new type. If
+     * the result failed, then we return {@code sameFailure}.
+     *
+     * @param function the function to apply to the item to create a new result
+     */
+    @NonNull
+    <TTo> Function<TFrom, Result<TTo>> thenApplyIfSucceeded(
+        @NonNull Function<? super TPrev, TTo> function);
+
+    /**
+     * Adds a {@link Function} to the behavior chain to map the item into a new result type. If the
+     * result failed, then we return {@code sameFailure}.
+     *
+     * @param function the function to apply to the item to create a new result
+     */
+    @NonNull
+    <TTo> FResult<TTo, TFrom> attemptApplyIfSucceeded(
+        @NonNull Function<? super TPrev, Result<TTo>> function);
+
+    /**
+     * Adds a {@link Function} to the end of the behavior chain to map the item into a new result
+     * type. If the result failed, then we return {@code sameFailure}.
+     *
+     * @param function the function to apply to the item to create a new result
+     */
+    @NonNull
+    <TTo> Function<TFrom, Result<TTo>> thenAttemptApplyIfSucceeded(
+        @NonNull Function<? super TPrev, Result<TTo>> function);
+
+    /**
+     * Adds a {@link Function} to the behavior chain to recover from the failure. If the result
+     * succeeded, then we return {@code get}.
+     *
+     * @param function the function to apply to the failure to recover
+     */
+    @NonNull
+    FItem<TPrev, TFrom> recover(@NonNull Function<? super Throwable, TPrev> function);
```

Comment:
I think both are useful. Sometimes you want to recover to a value rather than a result.